### ISSUE TITLE
Synapse v1.0.0rc1 significantly increases RAM usage

### ISSUE THREAD

**Issue by evilham** [matrix-org/synapse#5395]
### Description

My main Synapse server had been running happily with 4G RAM for some months, after upgrading to v1.0.0rc1 RAM usage increased a lot, leading to system crash.

After disabling presence (usual trick to reduce RAM usage), the RAM consumption went down enough for the server to be usable, but it is still out of the ordinary.

### Steps to reproduce

- upgrade to v1.0.0rc1
- see RAM usage increase
- beware of potential server crash

### Version information

- **Version**: v1.0.0rc1

- **Install method**: pip

Note: I waited until another user on #synapse-admins:matrix.org confirmed this.

**Comment by fredcy** [comment]
I'm running synapse v1.0.0rc1 on a vps with 8G RAM.  Not long after restarting synapse (less than an hour) I see it using about 2.8g of RES memory per the top command. Unfortunately, I don't have a baseline reading prior to upgrading to v1.0.0rc1.

**Comment by aaronraimist** [comment]
My RAM usage also spiked immediately after upgrading to 1.0.0rc1. Usually I hover around 2.1-2.8 GB used but it takes a while to get there. After upgrading 1.0.0rc1 my RAM usage was steadily climbing until it got to 3.7GB (out of 3.85GB total) and is currently sitting at 100% CPU on both cores.

**Comment by turt2live** [comment]
It's possible that the matrix.org outages are causing problems for people - how are extremities (#1760) looking?

**Comment by aaronraimist** [comment]
Extremities are ok. Highest is 5. (this is after Synapse OOMed and I stopped it from restarting but I assume it doesn't need to be running)

**Comment by aaronraimist** [comment]
Haven't setup grafana yet but this gets spammed in the logs as right memory usage is climbing rapidly. When I just restart my server it goes to about normal memory levels but it spikes as soon as I say something in a big room.

```
Jun 07 16:35:44 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:44,857 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-1171 - {PUT-O-1154} [yellowpen.net] Request failed: PUT matrix://yellowpen.net/_matrix/federation/v1/send/1559924920407: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:44 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:44,901 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-935 - {PUT-O-1070} [fff.global] Request failed: PUT matrix://fff.global/_matrix/federation/v1/send/1559924920323: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:44 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:44,944 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-1236 - {PUT-O-1174} [gottsnack.net] Request failed: PUT matrix://gottsnack.net/_matrix/federation/v1/send/1559924920427: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:44 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:44,972 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-790 - {PUT-O-1014} [ninetailed.net] Request failed: PUT matrix://ninetailed.net/_matrix/federation/v1/send/1559924920267: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:45 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:45,140 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-736 - {PUT-O-997} [vac.systemec.nl] Request failed: PUT matrix://vac.systemec.nl/_matrix/federation/v1/send/1559924920250: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:45 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:45,147 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-671 - {PUT-O-973} [m.searx.li] Request failed: PUT matrix://m.searx.li/_matrix/federation/v1/send/1559924920226: HttpResponseException("400: b'Bad Request'",)
Jun 07 16:35:45 ubuntu matrix-synapse[22132]: 2019-06-07 16:35:45,219 - synapse.http.matrixfederationclient - 517 - WARNING - federation_transaction_transmission_loop-696 - {PUT-O-984} [stephenson.cc] Request failed: PUT matrix://stephenson.cc/_matrix/federation/v1/send/1559924920237: HttpResponseException("400: b'Bad Request'",)
```

**Comment by evilham** [comment]
I have setup grafana, but the graphs are useless if I enable presence, since the server just crashes.
If presence is disabled, RAM usage is pretty much the same as before v1.0.0rc1 with presence enabled, but it does keep working.
Sadly, my graph is not very useful.

**Comment by evilham** [comment]
Just as an update: RAM usage did keep increasing until the host crashed even with presence disabled, apparently that just bought this server some time.

Started a `while true; do sv restart matrix-master; date; sleep 1800; done` on a screen meanwhile (it's terrible, does the job apparently).

**Comment by aaronraimist** [comment]
Adding our comments from `#synapse:matrix.org`:

For me, as long as I don't speak in large rooms the memory usage stays stable. As soon as I try to speak in `#synapse:matrix.org` or another large room then it crashes.

> aaron: wild guess: on this channel there are more servers without valid certs than on other channels, so that makes it worse here?

Seems to be correct.

**Comment by ingothierack** [comment]
Im on 0.99.5.2 current. Had activated "federation_verify_certificates" from first install a few weeks ago.
Have memory issues from the start.
Had now deactivated for around 2 hour, and have an stable memoy usage. After set to true, memory increase again.

![image](https://user-images.githubusercontent.com/11436841/59145900-864c2300-89ea-11e9-8dbf-903b2616d3f0.png)

10:20 restart with federation_verify_certificates: false. 12:20 restarted with true again. Memory grows shortly to 3.4 GB from 1.5 GB before. Maybe this is only more caching. But as you can see in the graph, it can groß up to 5.5 GB.
Maybe take a look in the verification process for memory issues.

**Comment by JJJollyjim** [comment]
I'm on 0.99.5.2, reproduced @ingothierack's results.

With a procedure of start the server then post in #synapse:matrix.org:
- `federation_verify_certificates: false` -> synapse process ends up around 700MB
- `federation_verify_certificates: true` -> synapse process grows >1.6GB and crashes the machine (which has always functioned okay on 2GB ram)

**Comment by JJJollyjim** [comment]
I added a stub to my synapse install that dumps a tracemalloc snapshot to a file when it receives a SIGUSR1.

This time, I used a procedure of start synapse, start typing in #synapse:matrix.org (thereby sending out typing notifications), watch the memory usage.

Again, with `federation_verify_certificates: false`, I end up stable (around 1GB), and it doesn't increase if I keep typing.

With `federation_verify_certificates: true`, the process keeps growing (much faster) until the machine dies.

I sent the SIGUSR1 to the first case once it had settled, and the second case when it was around 1.5GB and about to kill the machine.

Surprisingly, there is very little difference between the two!

The largest difference is +18.3MiB in `twisted/internet/defer.py`, which has a total of 25.5MiB of allocations... this does not at all explain the massive memory usage.

I see two possibilities:
- I don't know how to use tracemalloc (likely)
- All the significant memory usage is from non-python-controlled allocations: openssl or something perhaps?

I might see if I can figure out how to debug/prove the second case.

**Comment by JJJollyjim** [comment]
![image](https://user-images.githubusercontent.com/691552/59147659-8c75ea00-8a52-11e9-82e1-0230c0511501.png)

KDE heaptrack shows that 408MB (about the increase from don't-verify to verify) is allocated and never freed by openssl.

**Comment by evilham** [comment]
I can can confirm the issue is with the change of default for `federation_verify_certificates`, so it is not specific to v1.0.0rc1 but was present before.

**Comment by JJJollyjim** [comment]
I found this likely relevant issue: https://github.com/pyca/cryptography/issues/4833, which is fixed in the recently-released 2.7 (my version, provided by the latest matrix-provided deb, is 2.6)

**Comment by JJJollyjim** [comment]
(I say likely-relevant because the leaks are specifically from ASN.1 parsing, and the issue relates to X509 which is ASN.1-encoded)

**Comment by evilham** [comment]
@JJJollyjim: the server seeing issues on my side has `cryptography==2.7`, so I don't think it's *that* specific issue, though it is interesting to keep in mind.

**Comment by ara4n** [comment]
@JJJollyjim do you get backtraces in heaptrack to see where the leaky ASN.1 is being leaked from? Also, do you get the same behaviour if you upgrade to 2.7?

**Comment by JJJollyjim** [comment]
I do have the backtraces (of the C stack frames, not the python ones): all the leaks (actually about 800MB, now I've figured out how to read this) come from `_cffi_f_ssl_ctx_load_verify_locations` (i.e. the python binding to the openssl function `ssl_ctx_load_verify_locations`, which apparently is for loading CA certificates). That in turn is called by some unknown python function... `_PyMethodDef_RawFastCallKeywords`.

I could try and see if there is some allocation tracing thing that can also collection python function names from the stack frames. At a guess though, are we loading the CA store every time we do some sort of ssl thing, and keeping them around, instead of just having one shared copy?

**Comment by evilham** [comment]
Interesting. I've been trying a patch, it seems to work well so far. There is still increased RAM usage but not terribly so, there may still be room for improvement but it was written in 10 mins.

Basically, the way things are, each message sent over federation creates class and an instance of a class and there appears to be no caching whatsoever, this is the modified file:
https://matrix.unchat.cat/matrix_federation_agent.py
(Not on GH yet, sorry :-))

I'll see if I manage time-wise to add context caching and specifically look at the CA bits, but it'd be useful if someone could confirm that this is an improvement.

**Comment by JJJollyjim** [comment]
The patch definitely didn't fix the whole problem for me (can't say if it had any affect, as the memory usage is still at the point where the server crashes)

**Comment by evilham** [comment]
Thank you. I saw lower mem usage but not too much.
Care with testing of n a workstation, since the issue is triggered when federating with many servers.

**Comment by richvdh** [comment]
Thanks to everyone who has helped investigate this so far, and sorry we didn't catch it before putting out the RC.

> all the leaks (actually about 800MB, now I've figured out how to read this) come from _cffi_f_ssl_ctx_load_verify_locations (i.e. the python binding to the openssl function ssl_ctx_load_verify_locations, which apparently is for loading CA certificates).

This is very helpful indeed. Some background here:

Every time we start an outbound federation connection, [ClientTLSOptionsFactory.get_options](https://github.com/matrix-org/synapse/blob/v1.0.0rc1/synapse/crypto/context_factory.py#L141) is called. That in turn [calls](https://github.com/matrix-org/synapse/blob/v1.0.0rc1/synapse/crypto/context_factory.py#L155) `twisted.internet.ssl.CertificateOptions._makeContext`(https://github.com/twisted/twisted/blob/twisted-19.2.1/src/twisted/internet/_sslverify.py#L1682), which creates a fresh OpenSSL context and then calls `trustRoot._addCACertsToContext(ctx)`, which in this case loads the CA files from disk and attaches them to the new context.

In theory that should be fine, because the context (along with the loaded certs) gets thrown away when the connection is closed.

However, in practice, we are making several hundred outbound connections at once, and, thanks to http keepalive and connection pooling, the connections stay open for several minutes. What I therefore *suspect* is happening is that this is not a true leak, but simply massively increased memory usage thanks to attaching all of the CA certificates to each of the outbound connections. If Synapse lasted long enough to send all of the outbound requests and time out the connections, then we would see memory usage drop again. (The connection timeout is 120 seconds).

Intuitively it seems stupid that we are reading the certs from disk and storing them in a separate memory store for every single outbound connection; however, I'm not aware of a way to avoid this with the openssl interface. I don't think it is correct to share OpenSSL contexts between connections. Any thoughts or advice from those familiar with the OpenSSL API would be welcome.

**Comment by ara4n** [comment]
https://www.postgresql.org/message-id/E1bsROr-0002Z2-25%40gemulon.postgresql.org implies that it /is/ possible to share contexts between connections, albeit fiddly (which is why postgres stopped doing so)

**Comment by richvdh** [comment]
(https://github.com/matrix-org/synapse/issues/4673 was a bug that happened when we briefly tried sharing contexts: I'm not sure offhand if there is a better solution to that than having one-context-per-connection)

**Comment by ara4n** [comment]
http://openssl.6102.n7.nabble.com/Possibility-to-cache-ca-bundle-and-reuse-it-between-SSL-sessions-td51090.html looks to give most of the answers on how the API is meant to be used for this scenario.

**Comment by hawkowl** [comment]
Right now, you need one, because you set the info callback (which is how you do SNI) on the context, not the connection. https://github.com/twisted/twisted/pull/1128 will make it a little faster, though.

**Comment by JJJollyjim** [comment]
This node.js issue from 2011 also discusses the issue: https://paul.querna.org/articles/2011/04/05/openssl-memory-use/

They link this commit https://github.com/nodejs/node-v0.x-archive/commit/5c35dff4192b0e204ab4145b7f9dcdba6e76a93e which does the caching across contexts, and mention that they plan to make reuse contexts in future -- not sure if this ended up happening.

The other issue they discuss in the article - allocation of zlib buffers for TLS compression - is likely no longer relevant since TLS compression was turned off by default in openssl 1.1.0 (2016) to mitigate the CRIME attack

**Comment by richvdh** [comment]
Conclusion of discussion elsewhere is that you probably can share context objects, but have to do some hoop-jumping to get the SNI and cert verification done right.

Those interested can try playing with my `rav/shared_ssl_context` branch (https://github.com/matrix-org/synapse/tree/rav/shared_ssl_context). No promises that it won't blow up in your face.

**Comment by mvgorcum** [comment]
For some reason the extreme memory usage seems to have gone away for me, after running RC1 for a day.

**Comment by evilham** [comment]
Me and others have indeed seen a huge improvement `v1.0.0rc{2,3}`; thanks to everyone involved.

For the record: this improvement is also showing near 20% less RAM consumption on my biggest matrix server compared to `v0.99`, good thing this was addressed before `v1.0.0` :-).
